Refactor MCObjectHandle to allow MCButtonHandle, etc

livecodefraser opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/livecode/livecode/4430)

<!-- Reviewable:end -->

livecodefraser committed:
Add kObjectType constant to all MCObject classes

This value is used to implement safe casting via MCObjectCast.

livecodefraser committed:
Allow MCObjectCast to cast to MCObject*

This is necessary for some template-based magic elsewhere and is always
safe. As such, it doesn't impose any run-time checks.

livecodefraser committed:
Fix MCObjectCast to MCControl*

Audio clips and video clips are MCObject subclasses but are not
MCControls; the previous version allowed them to be cast incorrectly.

livecodefraser committed:
Refactor MCObjectHandle to allow MCButtonHandle, etc

Due to compiler limitations, classes like MCStackHandle won't convert
to MCObjectHandle naturally: this can be done in a type-safe manner
using some features in C++11 that are not available on all of our
supported platforms at this time.

montegoulding commented on engine/src/object.h:
```diff
@@ -1268,6 +1315,17 @@ class MCObject : public MCDLlist
 	bool getdeletionissuspended(void) const { return m_deletion_is_suspended; }
 	void setdeletionissuspended(bool value) { m_deletion_is_suspended = value; }
 
+    // Because C++11 support is universal amongst our supported platforms, auto-
```
@livecodefraser should there be a `not` before universal... otherwise I don't get this comment... but that could just be me ;-)

livecodefraser replied:
There should indeed! That'll teach me to not proof-read my comments :wink:

livecodefraser committed:
Update the friend declaration for MCMixinObjectHandle

GCC doesn't like the construction previously used (but Clang does) so
it has been updated to one that is acceptable to both.

livecodefraser committed:
Correct a comment about MCObjectHandles

A "not" was missing, completely reversing the meaning of the comment!

montegoulding commented:
Hmm... does this really need to target 8.2 as I was going to use it in #4392 

I guess we can pretty it up during the merge...

peter-b commented:
Reviewed 29 of 30 files at r1.
Review status: 29 of 30 files reviewed at latest revision, 4 unresolved discussions.

---

_[engine/src/object.h, line 33 [r2]](https://reviewable.io:443/reviews/livecode/livecode/4430#-KQVHY4HTRmjiVTl94Ue:-KQVHY4HTRmjiVTl94Uf:b-8y0zxs) ([raw file](https://github.com/livecode/livecode/blob/b90fc7cb4c122433ba7fca4a740ec13b0f44cfeb/engine/src/object.h#L33)):_

> ``` objc
> #include "native-layer.h"
> 
> // Disabled until C++11 support is available on all platforms
> ```

Please add a `// TODO(C++11)` here for greppability.

---

_[engine/src/object.h, line 230 [r2]](https://reviewable.io:443/reviews/livecode/livecode/4430#-KQVKTsK23Y_CVJwd5X5:-KQVKTsK23Y_CVJwd5X6:b-gfsskd) ([raw file](https://github.com/livecode/livecode/blob/b90fc7cb4c122433ba7fca4a740ec13b0f44cfeb/engine/src/object.h#L230)):_

> ``` objc
> 
>     // RAII handle class
>     class Handle;
> ```

BTW it might be worthwhile to add a comment explaining where to find the actual definition of this class.

---

_[engine/src/object.h, line 1588 [r2]](https://reviewable.io:443/reviews/livecode/livecode/4430#-KQVHSEML1s_uLQdN2Lm:-KQVHSEML1s_uLQdN2Ln:b-8y0zy5) ([raw file](https://github.com/livecode/livecode/blob/b90fc7cb4c122433ba7fca4a740ec13b0f44cfeb/engine/src/object.h#L1588)):_

> ``` objc
> 
>     // Conversion to handles to compatible types
>     // Currently disabled due to lack of required C++11 support on all platforms
> ```

Please add a `// TODO(C++11)` here for greppability!

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/livecode/livecode/4430)_

<!-- Sent from Reviewable.io -->

livecodefraser committed:
Merge branch 'develop' of https://github.com/runrev/livecode into mcfoohandle

Conflicts:
	engine/src/aclip.h
	engine/src/button.h
	engine/src/card.h
	engine/src/cpalette.h
	engine/src/eps.h
	engine/src/exec-interface2.cpp
	engine/src/field.h
	engine/src/graphic.h
	engine/src/group.h
	engine/src/image.h
	engine/src/magnify.h
	engine/src/mccontrol.h
	engine/src/object.cpp
	engine/src/object.h
	engine/src/player-legacy.h
	engine/src/player-platform.h
	engine/src/scrolbar.h
	engine/src/stack.h
	engine/src/vclip.h

livecodefraser committed:
Change MCButton::menu to an MCStackHandle

Instead of MCObjectHandle - this removes lots of needless casts.
